src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.cs

```diff
@@ -12,37 +14,194 @@ internal static partial class ClrThreadPool
+                    if (oldCounts == counts)
+                    {
+                        counts = newCounts;
+
+                        if(newCounts.maxWorking > oldCounts.maxWorking && ThreadPool.GetQueuedWorkItems().Any())
+                        {
+                            WorkerThread.MaybeAddWorkingWorker();
+                        }
+                    }
+                    else
+                    {
+                        counts = oldCounts;
+                    }
+                }
+                success = true;
             }
+            s_maxMinThreadLock.Release();
+            return success;
         }
 
         public static int GetMinThreads() => s_minThreads;
 
         public static bool SetMaxThreads(int threads)
         {
+            s_maxMinThreadLock.Acquire();
+            bool success;
             if (threads < s_minThreads || threads == 0)
             {
-                return false;
+                success = false;
             }
             else
             {
                 s_maxThreads = threads;
-                return true;
+
+                ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_counts);
+                while (counts.maxWorking > s_maxThreads)
+                {
+                    ThreadCounts newCounts = counts;
+                    newCounts.maxWorking = (short)s_maxThreads;
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_counts, newCounts, counts);
+                    if (oldCounts == counts)
+                    {
+                        counts = newCounts;
+                    }
+                    else
+                    {
+                        counts = oldCounts;
+                    }
+                }
+                success = true;
             }
+            s_maxMinThreadLock.Release();
+            return success;
         }
 
         public static int GetMaxThreads() => s_maxThreads;
+
+        internal static bool NotifyWorkItemComplete()
+        {
+            Interlocked.Increment(ref s_completionCount);
+            Volatile.Write(ref s_lastDequeueTime, Environment.TickCount);
+
+            bool shouldAdjustWorkers = ShouldAdjustMaxWorkersActive();
+            if(shouldAdjustWorkers && s_threadAdjustmentLock.TryAcquire())
+            {
+                AdjustMaxWorkersActive();
+                s_threadAdjustmentLock.Release();
+            }
+            return ShouldWorkerKeepRunning();
+        }
+
+        private static bool ShouldWorkerKeepRunning()
+        {
+            ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_counts);
+            while(true)
+            {
+                if(counts.numActive <= counts.numWorking)
+                {
+                    return true;
+                }
+
+                ThreadCounts newCounts = counts;
+                newCounts.numWorking--;
```

Comment:
`numActive` is the number of threads currently active in the thread pool. It is only decremented when the thread dies, which only happens when it times out.